src/app/AttributeAccessInterface.h

```diff
@@ -41,23 +42,135 @@
+        AttributePathIB::Builder attributePathIBBuilder     = attributeDataIBBuilder.CreatePath();
+        attributePathIBBuilder.Endpoint(mPath.mEndpointId).Cluster(mPath.mClusterId).Attribute(mPath.mAttributeId);
+
+        // Encode the list index field if we are encoding a single element in an list.
+        if (mCurrentEncodingListIndex != kInvalidListIndex)
+        {
+            attributePathIBBuilder.ListIndex(mCurrentEncodingListIndex);
+        }
+
+        ReturnErrorOnFailure(attributePathIBBuilder.EndOfAttributePathIB().GetError());
+
+        ReturnErrorOnFailure(
+            TagBoundEncoder(attributeDataIBBuilder.GetWriter(), TLV::ContextTag(to_underlying(AttributeDataIB::Tag::kData)))
+                .Encode(std::forward<Ts>(aArgs)...));
+
+        attributeDataIBBuilder.DataVersion(mDataVersion);
+
+        ReturnErrorOnFailure(attributeDataIBBuilder.EndOfAttributeDataIB().GetError());
+        return attributeReportIBBuilder.EndOfAttributeReportIB().GetError();
+    }
+
+    template <typename... Ts>
+    CHIP_ERROR EncodeListItem(Ts... aArgs)
+    {
+        if (mCurrentEncodingListIndex == kInvalidListIndex)
         {
+            if (mEncodeState.mCurrentEncodingListIndex == kInvalidListIndex)
+            {
+                // Spec 10.5.4.3.1, 10.5.4.6 (Replace a list w/ Multiple IBs)
+                // Put a empty array before encoding the first array element for list chunking.
+                AttributeReportIB::Builder attributeReportIBBuilder = mAttributeReportIBsBuilder.CreateAttributeReport();
+                AttributeDataIB::Builder attributeDataIBBuilder     = attributeReportIBBuilder.CreateAttributeData();
+                AttributePathIB::Builder attributePathIBBuilder     = attributeDataIBBuilder.CreatePath();
+
+                ReturnErrorOnFailure(attributePathIBBuilder.Endpoint(mPath.mEndpointId)
+                                         .Cluster(mPath.mClusterId)
+                                         .Attribute(mPath.mAttributeId)
+                                         .EndOfAttributePathIB()
+                                         .GetError());
+
+                ReturnErrorOnFailure(
+                    TagBoundEncoder(attributeDataIBBuilder.GetWriter(), TLV::ContextTag(to_underlying(AttributeDataIB::Tag::kData)))
+                        .EncodeList([](const TagBoundEncoder &) -> CHIP_ERROR { return CHIP_NO_ERROR; }));
+
+                attributeDataIBBuilder.DataVersion(mDataVersion).EndOfAttributeDataIB();
+                ReturnErrorOnFailure(attributeDataIBBuilder.GetError());
+                ReturnErrorOnFailure(attributeReportIBBuilder.EndOfAttributeReportIB().GetError());
+                mEncodeState.mCurrentEncodingListIndex = 0;
+            }
+            mCurrentEncodingListIndex = 0;
+        }
+
+        // We make Encode atomic here, so need to tell the caller do not rollback when we met any error during encoding.
```

Comment:
Why do we make Encode atomic in this case?

Reply:
Updated the comment
-- Because writer may contain unclosed containers or other partial data encoded due to error occurred, in this case, however, in the case of list chunking, we don't want the upper report engine rollback the writer (which will remove the whole attribute data encoded), so we make the Encode atomic and tell the report engine not to rollback in this case.